README.md
# News

**TODO: Add description**

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed as:

  1. Add news to your list of dependencies in `mix.exs`:

        def deps do
          [{:news, "~> 0.0.1"}]
        end

  2. Ensure news is started before your application:

        def application do
          [applications: [:news]]
        end



news.markdown
---
layout: post
title:  "Awesome Elixir"
date:   2016-04-09 00:00:11

---
## 2016-17

* [[news] multidimensional arrays in elixir](- http://blog.danielberkompas.com/2016/04/23/multidimensional-arrays-in-elixir.html) -  (2016-04-24)
* [Expectation  on elixir 1.3]( http://tuvistavie.com/tokyo.ex/#/) -    - Deprecation of imperative assignment
   - with improvements
   - Date datatypes
   - Plenty of other stuff (2016-04-24)
## 2016-16

* [Migrating from the elixir-lang-talk mailing list to Elixir Forum](http://elixirforum.com/t/elixir-lang-talk-is-migrating-to-this-forum/358) - The elixir-lang-talk mailing list has reached its limits. It doesn't provide a way to categorize contents, its search is limited, the lack of tools for maintenance makes it hard for us to manage it.  (2016-04-18)
* [What's new in Elixir 1.3](http://tuvistavie.com/tokyo.ex/#/) - Deprecation of imperative assignments, Date datatypes, a lot awesome stuff  (2016-04-19)
* [Why yet another lisp?(Video)](http://erlangcentral.org/8lu-with-robert-virding-erlang-concurrency-and-lfe-lisp-flavoured-erlang-2/#.Vxd9B5N969s) - LFE (Lisp Flavoured Erlang) has been designed to run efficiently on the Erlang VM. (2016-04-20)



test/news_test.exs
defmodule NewsTest do
  use ExUnit.Case
  doctest News

  test "the truth" do
    assert 1 + 1 == 2
  end
end


mix.exs
defmodule News.Mixfile do
  use Mix.Project

  def project do
    [app: :news,
     version: "0.0.1",
     elixir: "~> 1.2",
     build_embedded: Mix.env == :prod,
     start_permanent: Mix.env == :prod,
     deps: deps]
  end

  # Configuration for the OTP application
  #
  # Type "mix help compile.app" for more information
  def application do
    [mod: {News, [] },
     applications: [:logger, :tentacat, :timex]]
  end

  # Dependencies can be Hex packages:
  #
  #   {:mydep, "~> 0.3.0"}
  #
  # Or git/path repositories:
  #
  #   {:mydep, git: "https://github.com/elixir-lang/mydep.git", tag: "0.1.0"}
  #
  # Type "mix help deps" for more examples and options
  defp deps do
    [{:tentacat, git: "https://github.com/zhongwencool/tentacat.git", branch: "master"},
     {:timex, "~> 2.1.4"},
     {:exrm, "~> 1.0.3"},
    ]
  end

end



config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :news, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:news, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
config :news, github_access_token: System.get_env("GITHUB_ACCESS_TOKEN")
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/news/github.ex
defmodule News.Github do

  use Timex

  @client Tentacat.Client.new(%{access_token: Application.get_env(:news, :github_access_token)})
  @owner "elixir-news"
  @repo "elixir-news.github.io"
  @header "---\nlayout: post\ntitle:  \"Awesome Elixir\"\ndate:   2016-04-09 00:00:11\n\n---\n## "
  @news_file "_posts/2016-04-09-news.markdown"

  def update_news do
    issues = find_today_issues

    issues
    |> restructuring_issues
    |> upload_file

    close_issues(issues)
  end

  def restructuring_issues([]), do: :ok
  def restructuring_issues(issues) do
    new_news = joint_news(issues)
    file = Tentacat.Contents.find(@owner, @repo, @news_file, @client)
    content = file["content"] |> :base64.decode
    new_content = update_content(content, new_news)
    File.write("latest.news.markdown", new_content)
    %{sha: file["sha"], path: file["path"], message: commit_log, content: :base64.encode(new_content)}
  end

  defp update_content(content, new_news) do
    old_items = String.slice(content, 78..-1)
    week = week
    if String.starts_with?(old_items, week) do
      old_items = String.slice(old_items, 9..-1)
      @header <> week <> "\n\n" <> new_news <> old_items
    else
      @header <> week <> "\n\n" <> new_news <> "## "<> old_items
    end
  end

  defp joint_news(issues) do
    issues
    |> Enum.map(fn(issue) ->
      [url|text] = String.split(issue["body"], "\r\n\r\n")
      text = text |> List.to_string |> String.rstrip(?\n)
      date = String.slice(issue["created_at"], 0, 10)
      %{body: issue["body"],
        title: issue["title"],
        text: text,
        url: url,
        date: date,
        created_at: issue["created_at"]}
    end)
    |> Enum.sort_by(fn(x) -> x.created_at end)
    |> Enum.reduce("", fn(issue, acc) ->
      "* [#{issue.title}](#{issue.url}) - #{issue.text} (#{issue.date})\n" <> acc
    end)
  end

  def find_today_issues do
    {:ok, time} = Timex.DateTime.local |> Timex.beginning_of_day |> Timex.format("{ISO}")
    filters = %{state: "open", since: time}
    Tentacat.Issues.filter(@owner, @repo, filters, @client)
  end

  defp upload_file(body) do
    Tentacat.Contents.update(@owner, @repo, @news_file, body, @client)
  end

  def close_issues(issues) do
    Enum.each(issues, fn(issue) ->
      Tentacat.Issues.edit(@owner, @repo, issue["number"], %{state: "closed"}, @client)
    end)
  end

  defp week do
    {year, week} = :calendar.iso_week_number
    "#{year}-#{week}"
  end

  defp commit_log do
    Timex.DateTime.now("Asia/Shanghai") |> Timex.format!("%FT%T%:z", :strftime)
  end

end



lib/news/server.ex
defmodule News.Server do
  use GenServer

  @update_time [hours: 23, minutes: 50, seconds: 0]
  use Timex

  def start_link(state, opts \\ []) do
    GenServer.start_link(__MODULE__, state, opts)
  end

  def init(_) do
    timeout = remain_ms(@update_time)
    {:ok, {Timex.DateTime.local, :ok, timeout}, timeout}
  end

  def handle_info(:timeout, _state) do
    new_state =
      try  do
        News.Github.News.update
        New.Github.Awesome.update
        {Timex.DateTime.local, :ok, remain_ms(@update_time)}
      catch error ->
          {Timex.DateTime.local, error, remain_ms(@update_time)}
      end
    {:noreply, new_state, remain_ms(@update_time)}
  end

  def remain_ms(time) do
    update_datetime =
      "Asia/Shanghai"
      |> Timex.DateTime.today
      |> Timex.shift(time)
    now = Timex.DateTime.now
    diff_ms = Timex.diff(update_datetime, now, :seconds) * 1000
    if Timex.after?(now, update_datetime), do: 86400000 - diff_ms, else: diff_ms
  end

end



lib/news/awesome.ex
defmodule News.Github.Awesome do

  use Timex

  @client Tentacat.Client.new(%{access_token: Application.get_env(:news, :github_access_token)})
  @owner "elixir-news"
  @repo "elixir-news.github.io"
  @awesome_file "_posts/2016-04-09-awsome-elixir.markdown"
  @header """
  ---
  layout: post
  title:  "Awesome Elixir Collection"
  date:   2016-04-09 20:00:11
  ---

  Self-synchronizing from [awesome-elixir](https://github.com/h4cc/awesome-elixir) everyday.

  A curated list of amazingly awesome Elixir libraries, resources, and shiny things inspired by [awesome-php](https://github.com/ziadoz/awesome-php).

  """

  @source_owner "h4cc"
  @source_repo "awesome-elixir"
  @source_file "README.md"
  @last_source "latest_awesome_sha.md"

  def update do
    %{"content" => content, "sha" => sha} = fetch_source_file
    if need_update?(sha) do
      content
      |> restructuring_content
      |> upload_file
      File.write(@last_source, sha)
    end
  end

  def fetch_source_file do
    Tentacat.Contents.find(@source_owner, @source_repo, @source_file)
  end

  defp need_update?(sha) do
    case File.read(@last_source) do
      {:ok, ^sha} -> false
      _ -> true
    end
  end

  def restructuring_content(content) do
    file = Tentacat.Contents.find(@owner, @repo, @awesome_file, @client)
    new_content = content |> :base64.decode |> update_content
    %{sha: file["sha"], path: file["path"], message: commit_log, content: :base64.encode(new_content)}
  end

  defp update_content(content) do
    @header <> commit_log <> "\n\n" <> String.slice(content, 447..-1)
  end

  defp upload_file(body) do
    Tentacat.Contents.update(@owner, @repo, @awesome_file, body, @client)
  end

  defp commit_log do
    Timex.DateTime.now("Asia/Shanghai") |> Timex.format!("%FT%T%:z", :strftime)
  end

end



lib/news/news.ex
defmodule News.Github.News do

  use Timex

  @client Tentacat.Client.new(%{access_token: Application.get_env(:news, :github_access_token)})
  @owner "elixir-news"
  @repo "elixir-news.github.io"
  @header "---\nlayout: post\ntitle:  \"Awesome Elixir\"\ndate:   2016-04-09 00:00:11\n\n---\n## "
  @news_file "_posts/2016-04-09-news.markdown"

  def update do
    issues = find_today_issues

    issues
    |> restructuring_issues
    |> upload_file

    close_issues(issues)
  end

  def restructuring_issues([]), do: :ok
  def restructuring_issues(issues) do
    new_news = joint_news(issues)
    file = Tentacat.Contents.find(@owner, @repo, @news_file, @client)
    content = file["content"] |> :base64.decode
    new_content = update_content(content, new_news)
    File.write("latest.news.markdown", new_content)
    %{sha: file["sha"], path: file["path"], message: commit_log, content: :base64.encode(new_content)}
  end

  defp update_content(content, new_news) do
    old_items = String.slice(content, 78..-1)
    week = week
    if String.starts_with?(old_items, week) do
      old_items = String.slice(old_items, 9..-1)
      @header <> week <> "\n\n" <> new_news <> old_items
    else
      @header <> week <> "\n\n" <> new_news <> "## "<> old_items
    end
  end

  defp joint_news(issues) do
    issues
    |> Enum.map(fn(issue) ->
      [url|text] = String.split(issue["body"], "\r\n\r\n")
      text = text |> List.to_string |> String.rstrip(?\n)
      date = String.slice(issue["created_at"], 0, 10)
      %{body: issue["body"],
        title: issue["title"],
        text: text,
        url: url,
        date: date,
        created_at: issue["created_at"]}
    end)
    |> Enum.sort_by(fn(x) -> x.created_at end)
    |> Enum.reduce("", fn(issue, acc) ->
      "* [#{issue.title}](#{issue.url}) - #{issue.text} (#{issue.date})\n" <> acc
    end)
  end

  def find_today_issues do
    {:ok, time} = Timex.DateTime.local |> Timex.beginning_of_day |> Timex.format("{ISO}")
    filters = %{state: "open", since: time}
    Tentacat.Issues.filter(@owner, @repo, filters, @client)
  end

  defp upload_file(body) do
    Tentacat.Contents.update(@owner, @repo, @news_file, body, @client)
  end

  def close_issues(issues) do
    Enum.each(issues, fn(issue) ->
      Tentacat.Issues.edit(@owner, @repo, issue["number"], %{state: "closed"}, @client)
    end)
  end

  defp week do
    {year, week} = :calendar.iso_week_number
    "#{year}-#{week}"
  end

  defp commit_log do
    Timex.DateTime.now("Asia/Shanghai") |> Timex.format!("%FT%T%:z", :strftime)
  end

end

